src/analyze/analyze-unit-gdb-attach.c

```diff
@@ -0,0 +1,159 @@
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_free_ char *unit = NULL;
+        int r;
+
+        if (arg_transport != BUS_TRANSPORT_LOCAL)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Cannot spawn a unit shell for a remote service");
+
+        r = unit_name_mangle_with_suffix(argv[1], "as unit", UNIT_NAME_MANGLE_WARN, ".service", &unit);
+        if (r < 0)
+                return log_error_errno(r, "Failed to mangle name '%s': %m", argv[1]);
+
+        r = acquire_bus(&bus, /* use_full_bus= */ NULL);
+        if (r < 0)
+                return bus_log_connect_error(r, arg_transport, arg_runtime_scope);
+
+        _cleanup_free_ char *object = unit_dbus_path_from_name(unit);
+        if (!object)
+                return log_oom();
+
+        r = sd_bus_get_property(
+                        bus,
+                        "org.freedesktop.systemd1",
+                        object,
+                        "org.freedesktop.systemd1.Service",
+                        "MainPID",
+                        &error,
+                        &reply,
+                        "u");
+        if (r < 0)
+                return log_error_errno(r, "Failed to get the main PID of %s: %s", unit, bus_error_message(&error, r));
+
+        pid_t pid;
+        r = sd_bus_message_read(reply, "u", &pid);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read the main PID of %s from reply: %m", unit);
+
+        if (!arg_debugger) {
+                char *env_debugger;
+
+                env_debugger = getenv("SYSTEMD_DEBUGGER");
+                if (env_debugger)
+                        arg_debugger = env_debugger;
+                else
+                        arg_debugger = "gdb";
+        }
+
+        _cleanup_strv_free_ char **debugger_call = NULL;
+        r = strv_extend(&debugger_call, arg_debugger);
+        if (r < 0)
+                return log_oom();
+
+        _cleanup_free_ char *pid_str = NULL;
+        r = asprintf(&pid_str, "%d", pid);
+        if (r < 0)
+                return log_oom();
+
+        if (streq(arg_debugger, "gdb")) {
+                r = strv_extend_many(&debugger_call, "--pid", pid_str);
+                       if (r < 0)
+                               return log_oom();
+        }
+
+        if (streq(arg_debugger, "lldb")) {
+                r = strv_extend_many(&debugger_call, "--attach-pid", pid_str);
+                       if (r < 0)
+                               return log_oom();
+        }
+
+        if (arg_root) {
+                if (streq(arg_debugger, "gdb")) {
+                        r = strv_extend_many(&debugger_call, "--pid", pid_str);
+                               if (r < 0)
+                                       return log_oom();
+
+                        const char *sysroot_cmd;
+                        sysroot_cmd = strjoina("set sysroot ", arg_root);
```

Comment:
please do not use strjoina() with variables controlled by the user. i.e. use regular strjoin() here please, i.e. dynamic memory.